File: src/apps/mobile/src/api/catalog.ts
/**
 * Auto-cataloger API. Types mirror apps/api/app/schemas/catalog.py.
 */
import { Platform } from 'react-native';

import { api } from './client';

/** How a fact came to be known. Drives the provenance panel on the review step. */
export type FactSource = 'answered' | 'spoken' | 'unspecified';

export type QuestionType = 'chips' | 'chips_multi' | 'number' | 'text';

export type CraftQuestion = {
  id: string;
  /** Must match a ProductFacts field, or the answer is dropped on merge. */
  field: string;
  label: string;
  type: QuestionType;
  hint?: string | null;
  options?: string[] | null;
  unit?: string | null;
};

export type Craft = {
  id: string;
  label: string;
  icon: string;
  product_type: string;
  required: string[];
  questions: CraftQuestion[];
};

/** Every field optional by design — absent means "not specified", never guessed. */
export type ProductFacts = {
  product_type?: string | null;
  material?: string | null;
  technique?: string | null;
  colors?: string[];
  dimensions_cm?: string | null;
  weight_g?: number | null;
  piece_count?: number | null;
  care_instructions?: string | null;
  work_hours?: number | null;
  is_handmade?: boolean | null;
  customizations?: string[];
  occasion?: string | null;
};

export type TranscribeResult = {
  text: string;
  engine: string;
  language: string | null;
};

export type FollowUpQuestion = {
  field: string;
  question_en: string;
  question_native: string;
  audio_url: string | null;
};

export type ExtractFactsResult = {
  facts: ProductFacts;
  missing: string[];
  follow_ups: FollowUpQuestion[];
  sources: Record<string, FactSource>;
};

export type GeneratedListing = {
  title_en: string;
  title_hi: string;
  description_en: string;
  description_hi: string;
  seo_keywords: string[];
  bullet_points_en: string[];
  unspecified_fields: string[];
};

export type Violation = { claim: string; reason: string };

export type GenerateResult = {
  listing: GeneratedListing;
  /** Empty means the compliance pass found no unsourced claim. */
  violations: Violation[];
  /** True only when a check actually ran AND passed. */
  verified: boolean;
  /** False when no LLM was reachable, so no audit happened at all. */
  verification_ran: boolean;
};

/**
 * Extension for a recording's actual MIME type.
 *
 * This matters more than it looks: Whisper (both Groq's and the local one)
 * selects a decoder from the FILE EXTENSION, not the Content-Type. expo-audio
 * records webm/opus on web but m4a on native, so hardcoding ".m4a" hands the
 * decoder webm bytes under an m4a name and transcription silently returns
 * nothing usable.
 */
const AUDIO_EXT: Record<string, string> = {
  'audio/webm': 'webm',
  'audio/ogg': 'ogg',
  'audio/mp4': 'mp4',
  'audio/m4a': 'm4a',
  'audio/x-m4a': 'm4a',
  'audio/mpeg': 'mp3',
  'audio/wav': 'wav',
  'audio/x-wav': 'wav',
  'audio/3gpp': '3gp',
};

function extFor(mime: string, fallback: string): string {
  // Browsers report 'audio/webm;codecs=opus' — match on the base type.
  const base = mime.split(';')[0].trim().toLowerCase();
  return AUDIO_EXT[base] ?? fallback;
}

/**
 * Turn a local recording URI into something FormData will actually send,
 * named after what it actually is.
 *
 * Same platform split as mediaApi: React Native streams a {uri,name,type}
 * descriptor, the browser needs a real Blob or it posts "[object Object]".
 */
export async function toAudioFormValue(uri: string) {
  if (Platform.OS === 'web') {
    const blob = await (await fetch(uri)).blob();
    const type = blob.type || 'audio/webm';
    return {
      value: new File([blob], `note.${extFor(type, 'webm')}`, { type }),
      name: `note.${extFor(type, 'webm')}`,
    };
  }
  // expo-audio writes .m4a on iOS and .3gp on Android with LOW_QUALITY;
  // trust the URI's own extension rather than assuming.
  const uriExt = uri.split('?')[0].split('.').pop()?.toLowerCase();
  const ext = uriExt && uriExt.length <= 4 ? uriExt : 'm4a';
  const name = `note.${ext}`;
  const type = Object.entries(AUDIO_EXT).find(([, e]) => e === ext)?.[0] ?? 'audio/m4a';
  return { value: { uri, name, type } as unknown as Blob, name };
}

export const catalogApi = {
  crafts: () => api.get<Craft[]>('catalog/crafts'),

  /** `lang` may be 'auto' to let Whisper detect the language itself. */
  transcribe: async (uri: string, lang = 'hi') => {
    const { value } = await toAudioFormValue(uri);
    const form = new FormData();
    form.append('file', value);
    form.append('lang', lang);
    return api.upload<TranscribeResult>('catalog/transcribe', form);
  },

  extractFacts: (body: {
    transcript: string;
    lang?: string;
    craft_id?: string | null;
    existing_facts?: ProductFacts;
  }) => api.post<ExtractFactsResult>('catalog/extract-facts', body),

  generate: (body: {
    facts: ProductFacts;
    craft_id?: string | null;
    craft_class?: string | null;
    artisan_words?: string | null;
    title_override?: string | null;
    tone?: 'marketplace' | 'premium' | 'b2b';
  }) => api.post<GenerateResult>('catalog/generate', body),
};

File: src/apps/api/app/schemas/catalog.py
"""
Catalog schemas -- the anti-hallucination contract in type form.

Every ProductFacts field is Optional on purpose. A None is rendered downstream
as "not specified"; it is NEVER filled in by a model. That single rule is what
stops the failure Features.docx names explicitly: writing "100% pure silk" over
a blend is a legal problem and a return, not a copywriting flourish.

Deviation from PRD.md:1958 -- the PRD declares `REQUIRED: tuple` as an attribute
on the model itself. Under Pydantic v2 that becomes a real field (which is why
the PRD's own code has to `model_dump(exclude={"REQUIRED"})`). Required fields
are per-craft here anyway, so they live in app/data/crafts.py and are looked up,
not carried on the instance.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.data.crafts import required_fields

# How a fact came to be known. Surfaced in the app as provenance, so the artisan
# can see that nothing was invented on her behalf.
FactSource = Literal["answered", "spoken", "unspecified"]


class ProductFacts(BaseModel):
    """Confirmed facts only. Absent means unknown, never means 'assume'."""

    product_type: str | None = Field(None, description="saree, pot, toy, painting")
    material: str | None = Field(None, description="exact material as stated")
    technique: str | None = None
    colors: list[str] = Field(default_factory=list)
    dimensions_cm: str | None = None
    weight_g: int | None = None
    piece_count: int | None = None
    care_instructions: str | None = None
    work_hours: float | None = Field(None, description="hours to make ONE unit")
    is_handmade: bool | None = None
    customizations: list[str] = Field(default_factory=list)
    occasion: str | None = None

    @field_validator("colors", "customizations", mode="before")
    @classmethod
    def _null_means_empty(cls, v):
        """Accept null for the list fields.

        The extraction prompt tells the model to leave anything unmentioned as
        null, and it obeys -- including for lists. Rejecting `"colors": null`
        made Pydantic throw away an otherwise perfect extraction, so a fully
        correct answer was being discarded over a type technicality. A single
        string is also tolerated: models occasionally emit "blue" for a
        one-item list.
        """
        if v is None:
            return []
        if isinstance(v, str):
            return [v] if v.strip() else []
        return v

    def missing_required(self, craft_id: str | None = None) -> list[str]:
        return [
            f for f in required_fields(craft_id)
            if getattr(self, f, None) in (None, "", [])
        ]

    def confirmed(self) -> dict:
        """Only what is actually known. This is all the generator ever sees."""
        return {k: v for k, v in self.model_dump().items() if v not in (None, "", [])}

    def unknown(self) -> list[str]:
        """The explicit forbidden list handed to the generator."""
        return [k for k, v in self.model_dump().items() if v in (None, "", [])]

    def merge(self, other: "ProductFacts") -> "ProductFacts":
        """Fold `other` in WITHOUT overwriting anything already known.

        Direction matters and is the whole grounding argument: questionnaire
        answers are stated directly by the artisan, while spoken facts are
        inferred by a model from a transcript. When they disagree, the stated
        answer wins. Call as `answered.merge(extracted)`.
        """
        merged = self.model_dump()
        for key, value in other.model_dump().items():
            if merged.get(key) in (None, "", []) and value not in (None, "", []):
                merged[key] = value
        return ProductFacts(**merged)


class FollowUpQuestion(BaseModel):
    field: str
    question_en: str
    # Kept for the app to speak in the artisan's language. There is no
    # server-side TTS any more -- expo-speech renders this on the device.
    question_native: str
    audio_url: str | None = None


class ExtractFactsRequest(BaseModel):
    transcript: str
    lang: str = "hi"
    craft_id: str | None = None
    existing_facts: dict = Field(default_factory=dict)


class ExtractFactsResponse(BaseModel):
    facts: ProductFacts
    missing: list[str]
    follow_ups: list[FollowUpQuestion]
    # Answers the server could not read -- a spoken phrase in a numeric box,
    # typically. Reported rather than swallowed so the app can ask again
    # instead of letting the field quietly vanish from the listing.
    discarded: list[str] = Field(default_factory=list)
    # field -> how we know it. Drives the provenance panel.
    sources: dict[str, FactSource] = Field(default_factory=dict)


class GeneratedListing(BaseModel):
    title_en: str
    title_hi: str
    description_en: str
    description_hi: str
    seo_keywords: list[str]
    bullet_points_en: list[str]
    unspecified_fields: list[str] = Field(default_factory=list)


class GenerateRequest(BaseModel):
    facts: ProductFacts
    craft_id: str | None = None
    craft_class: str | None = None
    gi_tag: str | None = None
    artisan_name: str | None = None
    village: str | None = None
    # The artisan's own words, if she recorded a free description. Used as raw
    # material for voice and tone -- never as a source of new facts, which is
    # why extraction is a separate call.
    artisan_words: str | None = None
    # Explicit title edits are applied verbatim after generation so a voice
    # instruction such as "change the title to ..." is not paraphrased.
    title_override: str | None = Field(default=None, max_length=300)
    tone: Literal["marketplace", "premium", "b2b"] = "marketplace"


class Violation(BaseModel):
    claim: str
    reason: str


class GenerateResponse(BaseModel):
    listing: GeneratedListing
    # Empty means the compliance pass found no unsourced claim. The app shows
    # this as the "nothing invented" badge, so it must be honest even when the
    # answer is inconvenient.
    violations: list[Violation] = Field(default_factory=list)
    # True only when a check actually ran AND passed. When the LLM is
    # unreachable the copy comes from the deterministic template, which is
    # grounded by construction but has not been *audited* -- claiming it was
    # would be exactly the kind of unearned confidence this feature exists to
    # prevent.
    verified: bool = True
    verification_ran: bool = True


class TranscribeResponse(BaseModel):
    text: str
    engine: str
    language: str | None = None


class CraftQuestion(BaseModel):
    id: str
    field: str
    label: str
    type: str
    hint: str | None = None
    options: list[str] | None = None
    unit: str | None = None


class CraftOut(BaseModel):
    id: str
    label: str
    icon: str
    product_type: str
    required: list[str]
    questions: list[CraftQuestion]


__all__ = [
    "CraftOut",
    "CraftQuestion",
    "ExtractFactsRequest",
    "ExtractFactsResponse",
    "FollowUpQuestion",
    "GenerateRequest",
    "GenerateResponse",
    "GeneratedListing",
    "ProductFacts",
    "TranscribeResponse",
    "Violation",
]
